src/modules/job-manager/plugins/perilog.c

```diff
@@ -617,35 +617,49 @@ static bool perilog_per_rank (struct perilog_proc *proc)
     return perilog_config.epilog->per_rank;
 }
 
+static void proc_drain_and_finish (struct perilog_proc *proc,
+                                   bool drain_failed,
+                                   bool drain_active)
+{
+
+    if (drain_failed || drain_active) {
+        /* Drain the set of ranks that failed the prolog/epilog. If the
+         * drain RPC is successful, then wait for the response before
+         * emitting the "prolog/epilog-finish" event. O/w, resources could
+         * be freed and handed out to new jobs before they are drained.
+         */
+        if ((proc->drain_f = proc_drain_ranks (proc,
+                                               drain_failed,
+                                               drain_active))
+                && flux_future_then (proc->drain_f,
+                    -1.,
+                    drain_failed_cb,
+                    proc) == 0)
+            return;
+
+        /* O/w, drain RPC failed, report error and fall through so finish
+         * event is still emitted.
+         */
+        flux_log_error (flux_jobtap_get_flux (proc->p),
+                        "%s: failed to drain %s failed ranks",
+                        idf58 (proc->id),
+                        perilog_proc_name (proc));
```

Comment:
Minor:  `proc_drain_ranks()` also logs on error.  Since all it and this are doing is preparing and sending a request, it seems like one log message should suffice.

Reply:
Then no error would be logged if `flux_future_then()` fails. Rare enough not to matter? Eh, probably. I'll remove it.

Reply:
Eh, I wasn't advocating for no logging on `flux_future_then()` failure.  Just that maybe the logging could be consolidated somehow.  Maybe not worth changing it.